aiida_vasp/workchains/vasp.py

```diff
@@ -273,3 +297,95 @@ def on_except(self, exc_info):
+        the output. The same problems may also be the hidden root of a calculation failure. For that reason,
+        after verifying that the calculation ran, regardless of its calculation state, we perform some sanity
+        checks. However, make sure that this workchain is general. Calculation plugin specific sanity checks
+        should go into the childs..
+
+        Note that this is the "final" check, most of the potential problems should have been handled and
+        fixed in restart by error handlers.
+        """
+
+        misc = node.outputs.misc.get_dict()
+        if 'run_status' not in 'misc':
+            self.report('`run_status` is not found in misc - cannot verify the integrity of the child calcualtio.')
+            return ProcessHandlerReport(exit_code=self.exit_codes.ERROR_UNKNOWN)
+
+        run_status = misc['run_status']
+
+        # Check if the calculation is indeed finished
+        if not run_status.get('finished'):
+            self.report(f'The child calcualtion {node} did not reach the end of execution.')
+            return ProcessHandlerReport(exit_code=self.exit_codes.ERROR_CALCUALTION_NOT_FINISHED)
+
+        # Check that the electronic structure is converged
+        if not run_status.get('electronic_converged'):
+            self.report(f'The child calcualtion {node} did not have converged electronic structure.')
+            return ProcessHandlerReport(exit_code=self.exit_codes.ERROR_ELECTRONIC_STRUCTURE_NOT_CONVERGED)
+
+        return None
+
+    @process_handler(priority=100)
+    def _handle_unfinished_calculation(self, node):
+        """
+        Handled the problem such that the calculation is not finished, e.g. did not reach the
+        end of execution.
+
+        If WAVECAR exists, just resubmit the calculation with the restart folder.
+
+        If it is a geometry optimisation, attempt to restart with output structure + WAVECAR.
+        """
+
+        misc = node.outputs.misc.get_dict()
+        run_status = misc['run_status']
+
+        # No need to act if the process is finished
+        if run_status.get('finished'):
+            return None
+
+        # Check it is a geometry optimisation
+        incar = self.ctx.inputs.parameters
+        if incar.get('nsw', -1) > 0:
+            if 'structure' not in node.outputs.structure:
+                self.report('Performing a geometry optimisation but the output structure is not found.')
+                return ProcessHandlerReport(do_break=True, exit_code=self.exit_codes.ERROR_MANUAL_INTERVENTION_NEEDED)
+            self.report('Continuing geometry optimisation using the last geometry.')
+            self.ctx.inputs.structure = node.outputs.structure
+            self._attach_wavecar(node)
+        else:
+            # Single point calculation - attach the wavecar if possible
+            has_wavecar = self._attach_wavecar(node)
+            if not has_wavecar:
+                # If there is no WAVECAR there is not much we can do.....
```

Comment:
Yeah but system issues are not trivial to detects universally and relatively rare 🤞 , reaching the wall-time limit is the most likely cause I think. Without the WAVECAR there is not much point to restart a single-point calculation.

Reply:
Not necessarily. Since we do not have the scheduler parsing nailed in `aiida_core` yet, we would have to also assume such failures as "system failures", or at least, VASP did not start. Often a new restart works, sometimes because it distributes the job to different nodes after failed nodes have been tagged for failure the first time. Etc.

But the restart workchain had his functionality. I have not yet checked if this is still the case. Default was 5 dumb restarts if we could not detect what was going wrong. I think such an approach is still pretty nice to have, since the user does not have to lift a finger.

Reply:
I will update it such that we retry it once, although this may risk wasting the resources 💰. Also, I realised that for unfinished calculations that are not hybrid ionic relaxation, there won't be any WAVECAR at all....

Reply:
If it does not a start at all, hardly any resources are wasted. However, the HPC center might not like the aggressive job submission policy :)